src/app/services/product.service.ts

```diff
@@ -1,28 +1,74 @@
 import { Injectable } from '@angular/core';
 import { AngularFireDatabase } from '@angular/fire/database';
+import { BehaviorSubject } from 'rxjs';
+import { FilterForm } from '../models/filter-form.model';
 import { Product } from '../models/product.model';
 
 @Injectable({
   providedIn: 'root'
 })
 export class ProductService {
-  public products: Product[];
+  private originProducts: Product[];
+  public products = new BehaviorSubject([]);
   constructor(private db: AngularFireDatabase) { }
 
-  getProducts(): Promise<any[]> {
+  public getProducts(): Promise<any[]> {
     return new Promise((resolve) => {
       this.db.database.ref(`products`).once('value').then(snapshot => {
-        this.products = snapshot.val();
+        this.originProducts = snapshot.val();
+        this.products.next(snapshot.val());
         resolve(snapshot.val());
       });
     });
   }
 
-  getProductById(id: number): Promise<any> {
+  public getProductById(id: number): Promise<any> {
     return new Promise((resolve) => {
       this.db.database.ref(`products/${id}`).once('value').then(snapshot => {
         resolve(snapshot.val());
       });
     });
   }
+
+  public sorting(condition): void {
+    const sortedProducts = this.products.getValue().sort((a, b) => {
+      return eval(`${condition.operator === '-' ? 'a.' : 'b.'}${condition.value}
+                      -
+                      ${condition.operator === '+' ? 'a.' : 'b.'}${condition.value}`);
+    });
+    this.products.next(sortedProducts);
+  }
+
+  public filter(form: FilterForm): void {
+    let c = [];
```

Comment:
Create semantic variables!

Reply:
I agree with @iyarymovych-reenbit, I dont understand how this code works

Reply:
fixed